gadget/filesystemimage_test.go

```diff
@@ -0,0 +1,375 @@
+	c.Assert(err, IsNil)
+	c.Assert(matches, HasLen, 0)
+}
+
+func (s *filesystemImageTestSuite) TestPostStageOptional(c *C) {
+	var mkfsCalled bool
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		mkfsCalled = true
+		return nil
+	}
+
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, s.psTrivial)
+
+	err = fiw.Write(img, nil)
+	c.Assert(err, IsNil)
+	c.Assert(mkfsCalled, Equals, true)
+}
+
+func (s *filesystemImageTestSuite) TestChecksImage(c *C) {
+	cb := func(rootDir string, cbPs *gadget.PositionedStructure) error {
+		return errors.New("unexpected call")
+	}
+
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("unexpected mkfs call")
+	}
+
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := filepath.Join(s.dir, "img")
+
+	// no image file
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, "cannot stat image file: .*/img: no such file or directory")
+
+	// image file smaller than expected
+	makeSizedFile(c, img, s.psTrivial.Size/2, nil)
+
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, fmt.Sprintf("size of image file %v is different from declared structure size %v", s.psTrivial.Size/2, s.psTrivial.Size))
+}
+
+func (s *filesystemImageTestSuite) TestPostStageError(c *C) {
+	cb := func(rootDir string, cbPs *gadget.PositionedStructure) error {
+		return errors.New("post stage exploded")
+	}
+
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("unexpected mkfs call")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, s.psTrivial)
+
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, "post stage callback failed: post stage exploded")
+}
+
+func (s *filesystemImageTestSuite) TestMkfsError(c *C) {
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("will not mkfs")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, s.psTrivial)
+
+	err = fiw.Write(img, nil)
+	c.Assert(err, ErrorMatches, `cannot create "happyfs" filesystem: will not mkfs`)
+}
+
+func (s *filesystemImageTestSuite) TestFilesystemExtraCheckError(c *C) {
+	ps := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Filesystem: "happyfs",
```

Comment:
Would it be easier to just use "ext4" here and then you can skip the mocking of mkfsHappyFs and MockMkfsHandlers? AFAICT you only need it because NewFilesystemImageWriter is also doing validation of the handler. And the handler is changed before the fiw.Write() call so that should be ok?

Reply:
Switcher to suite setup level mocks.